### DISCUSSION TITLE
New extension point for Authz service

### DISCUSSION THREAD

**Discussion by scandinave** [eclipse-edc/Connector#4575] [Ideas]
Hi, 

Currently there is an AuthenticationService extension point to handle authentication on the ManagementAPI. We would like to secure the API with an OAuth2 provider. 

The AuthenticationServiceExtension point does not let extension access the whole request and the target endpoint, only the header so it's not possible to make good decision for authorization. We propose to update the AuthenticationService for this use case ou create a new AuthzService extension point that is registered in `ManagementApiConfigurationExtension` after the AuthenticationExtensionPoint and that can pass the full request to the extension implementation.  

Is this something that is possible to contribute?

**Comment by paullatzelsperger** [comment]
hi @scandinave. what information from the request would you use to make the authz decision, and how? 

Please read our [adoption guidelines](https://eclipse-edc.github.io/documentation/adoption/#get-adopted-in-edc-core) in case you haven't.

**Reply by scandinave** [reply]
We need to have access to the endpoint requested and eventually to the request body to provide claims to the Authorization provider.

This implies to validate statically the Bearer token with scope and permission for the target endpoint or dynamically by calling the authorization provider with this bearer token.

This is just a preliminary question to known if there haven't already a decision made to not support this use case. I will, of course,  submit a proper issues if this is something that could be considered

**Reply by paullatzelsperger** [reply]
You are effectively talking about RBAC. EDC does not have a concept of a "user" or a "role".

The EDC project purposely opted **against** providing any elaborate authz mechanisms for the Management API, and **for** providing an easy extension mechanism, so every project can implement this as they need. 

Our reasoning behind this is, that many adopters have many different requirements, and many different understandings of roles and users etc., so implementing (and maintaining!) this in a generic way is not appealing to us.

[edit]: if you have a workable solution though, you could have it listed as [friend of EDC](https://eclipse-edc.github.io/documentation/adoption/#get-referenced-as-friend)

**Reply by scandinave** [reply]
I understand this. The goal is only to provide an ExtensionPoint with sufficient information to allow implementation like our to handle this case.

**Reply by paullatzelsperger** [reply]
nothing is needed from EDC side here: simply implement a `ContainerRequestFilter`, register it to all Management API controllers and implement your custom logic there.

**Reply by scandinave** [reply]
Yes thanks for your time

**Reply by paullatzelsperger** [reply]
can I close this discussion as solved then?

**Reply by scandinave** [reply]
Yes please
